verification/policy.go

```diff
@@ -0,0 +1,130 @@
+	// TrustStore this policy stament uses
+	TrustStore string `json:"trustStore,omitempty"`
+	// TrustedIdentities this policy statement pins
+	TrustedIdentities []string `json:"trustedIdentities,omitempty"`
+}
+
+func isPresent(val string, values []string) bool {
+	for _, v := range values {
+		if v == val {
+			return true
+		}
+	}
+	return false
+}
+
+// ValidatePolicyDocument validates a policy document according to it's version's rule set.
+// if any rule is violated, returns an error
+func ValidatePolicyDocument(policyDoc *PolicyDocument) error {
+	// Constants
+	wildcard := "*"
+	supportedPolicyVersions := []string{"1.0"}
+	supportedVerificationPresets := []string{"strict", "permissive", "audit", "skip"}
+	supportedTrustStorePrefixes := []string{"ca", "signingservice"}
+
+	// Validate Version
+	if !isPresent(policyDoc.Version, supportedPolicyVersions) {
+		return fmt.Errorf("version %q is not supported", policyDoc.Version)
+	}
+
+	// Validate the policy according to 1.0 rules
+	if len(policyDoc.TrustPolicies) == 0 {
+		return errors.New("trust policy document can not have zero policy statements")
+	}
+	policyStatementNameCount := make(map[string]int)
+	registryScopeCount := make(map[string]int)
+	for _, statement := range policyDoc.TrustPolicies {
+
+		// Verify statement name is valid
+		if statement.Name == "" {
+			return errors.New("policy statement is missing a name")
+		}
+		policyStatementNameCount[statement.Name]++
+
+		// Verify registry scopes are valid
+		if len(statement.RegistryScopes) == 0 {
+			return fmt.Errorf("policy statement %q has zero registry scopes", statement.Name)
+		}
+		if len(statement.RegistryScopes) > 1 && isPresent(wildcard, statement.RegistryScopes) {
+			return errors.New("wildcard registry scope can not be shared with other registries")
+		}
+		for _, scope := range statement.RegistryScopes {
+			registryScopeCount[scope]++
+		}
+
+		// Verify signature verification preset is valid
+		if !isPresent(statement.SignatureVerification, supportedVerificationPresets) {
+			return fmt.Errorf("signatureVerification %q is not supported", statement.SignatureVerification)
+		}
+
+		// Any signature verification other than "skip" needs a trust store
+		if statement.SignatureVerification != "skip" && (statement.TrustStore == "" || len(statement.TrustedIdentities) == 0) {
+			return fmt.Errorf("%q is either missing a trust store or trusted identities", statement.Name)
+		}
+
+		// Verify trust store type is valid if it is present (trust store is optional for "skip" signature verification)
+		if statement.TrustStore != "" {
+			i := strings.Index(statement.TrustStore, ":")
+			if i < 0 || !isPresent(statement.TrustStore[:i], supportedTrustStorePrefixes) {
+				return fmt.Errorf("%q has a trust store with an unsupported trust store type", statement.Name)
+			}
+		}
+
+		// If there are trusted identities, verify they are not empty
+		for _, identity := range statement.TrustedIdentities {
```

Comment:
Would be better to move trusted identities validation into its own method. Validation for trusted identity `x509.subject` and valid and minimum RDNs is missing.

https://github.com/notaryproject/notaryproject/blob/main/trust-store-trust-policy-specification.md#trusted-identities-constraints

Reply:
Will address this in a separate PR